server/lib/two-factor-authentication/lib.ts

```diff
@@ -0,0 +1,168 @@
+import { ApolloError } from 'apollo-server-errors';
+import { Request } from 'express';
+import { isNil } from 'lodash';
+
+import User from '../../models/User';
+import cache from '../cache';
+
+import totp from './totp';
+
+const DEFAULT_TWO_FACTOR_AUTH_SESSION_DURATION = 600; // 10min
+
+type ValidateRequestOptions = {
+  // require user configured 2FA
+  requireTwoFactorAuthEnabled?: boolean;
+  // always ask for a token when using 2FA
+  alwaysAskForToken?: boolean;
+  // duration which we wont require a token after a successful use
+  sessionDuration?: number;
+  // identifier for the session
+  sessionKey?: (() => string) | string;
+};
+
+export enum TwoFactorMethod {
+  TOTP = 'totp',
+}
+
+export const TwoFactorAuthenticationHeader = 'x-two-factor-authentication';
+
+export const SupportedTwoFactorMethods = [TwoFactorMethod.TOTP];
+
+export type Token = {
+  type: TwoFactorMethod;
+  code: string;
+};
+
+export interface TwoFactorAuthProvider {
+  validateToken(user: typeof User, token: Token): Promise<void>;
+}
+
+export const providers: { [method in TwoFactorMethod]: TwoFactorAuthProvider } = {
+  [TwoFactorMethod.TOTP]: totp,
+};
+
+function getTwoFactorAuthTokenFromRequest(req: Request): Token {
+  const header = req.get(TwoFactorAuthenticationHeader);
+  if (!header) {
+    return null;
+  }
+
+  const parts = header.split(' ');
+  const type = parts[0] as TwoFactorMethod;
+  const code = parts[1];
```

Comment:
I haven't actually tested (would be great to have some unit tests to cover the actual errors for this one) but I think we can be a bit more strict and do some pre-validation here in case people try to fool with the `x-two-factor-authentication` header. Some examples:

> { x-two-factor-authentication: ' ' } => { type: '', code: '' }
> { x-two-factor-authentication: 'totp' } => { type: 'totp', code: undefined }

```suggestion
  const parts = header.split(' ');
  const type = parts[0] as TwoFactorMethod;
  const code = parts[1];
  if (!type || !code) {
    throw new Error(`Malformed 2FA token header`);
  }
```